pkg/credentialprovider/aws/aws_credentials.go

```diff
@@ -61,172 +103,188 @@ type ecrTokenGetter struct {
-// Client code will call Provider.LazyProvide() at image pulling time.
-func (p *lazyEcrProvider) Provide() credentialprovider.DockerConfig {
-	entry := credentialprovider.DockerConfigEntry{
-		Provider: p,
-	}
+func (p *ecrProvider) getFromECR(parsed *parsedURL) credentialprovider.DockerConfig {
+	klog.Infof("Getting credentials from ECR for %v", parsed.host)
 	cfg := credentialprovider.DockerConfig{}
-	cfg[p.regionURL] = entry
+	if p.getter == nil {
+		p.getter = newGetter(parsed.region)
+	}
+
+	params := &ecr.GetAuthorizationTokenInput{RegistryIds: []*string{aws.String(parsed.registryID)}}
+	output, err := p.getter.GetAuthorizationToken(params)
+	if err != nil {
+		klog.Errorf("while requesting ECR authorization token %v", err)
+		return cfg
+	}
+	if output == nil {
+		klog.Errorf("Got back no ECR token")
+		return cfg
+	}
+
+	data := output.AuthorizationData[0]
+	if data.ProxyEndpoint != nil &&
+		data.AuthorizationToken != nil {
+		decodedToken, err := base64.StdEncoding.DecodeString(aws.StringValue(data.AuthorizationToken))
+		if err != nil {
+			klog.Errorf("while decoding token for endpoint %v %v", data.ProxyEndpoint, err)
+			return cfg
+		}
+		parts := strings.SplitN(string(decodedToken), ":", 2)
+		user := parts[0]
+		password := parts[1]
+		creds := credentialprovider.DockerConfigEntry{
+			Username: user,
+			Password: password,
+			// ECR doesn't care and Docker is about to obsolete it
+			Email: "not@val.id",
+		}
+		entry := cacheEntry{
+			expiresAt:   *data.ExpiresAt,
+			credentials: creds,
+			host:        parsed.host,
+		}
+		if err := p.cache.Add(entry); err != nil {
+			klog.Errorf("while adding entry to cache %v", err)
+			return cfg
+		}
+		cfg[entry.host] = entry.credentials
+	}
 	return cfg
 }
 
-func newEcrProvider(region string, getter tokenGetter) *ecrProvider {
+/*PROVIDER*/
+// ecrProvider is a DockerConfigProvider that gets and refreshes tokens
+// from AWS to access ECR.
+type ecrProvider struct {
+	registryURLTemplate string
+	cache               cache.Store
+	getter              tokenGetter
+}
+
+var _ credentialprovider.DockerConfigProvider = &ecrProvider{}
+
+func newECRProvider(template string, getter tokenGetter) *ecrProvider {
 	return &ecrProvider{
-		region:    region,
-		regionURL: registryURL(region),
-		getter:    getter,
+		registryURLTemplate: template,
+		// cache:               cache.NewTTLStore(stringKeyFunc, 1*time.Hour),
+		cache:  cache.NewStoreWithPolicy(stringKeyFunc, &ecrExpirationPolicy{}),
+		getter: getter,
 	}
 }
 
+// init registers a credential provider for the specified region.
+// It creates a provider for each registryURLTemplate, in order to
+// support cross-region ECR access.
+func init() {
+	credentialprovider.RegisterCredentialProvider("aws-ecr-partition-standard",
+		newECRProvider(registryURLTemplateStandard, nil))
+
+	credentialprovider.RegisterCredentialProvider("aws-ecr-partition-china",
+		newECRProvider(registryURLTemplateChina, nil))
+}
```

Comment:
Kind of a nitpick but would you move `init` to the top just under the constants so it's easy to see that there's init behavior as soon as you open the file?

Reply:
I'll think about it